rasa/nlu/utils/spacy_utils.py

```diff
@@ -129,18 +130,61 @@ def get_text(self, example, attribute):
 
         return self.preprocess_text(example.get(attribute))
 
+    @staticmethod
+    def merge_content_lists(freezed_indices, doc_lists):
+        dct = dict(freezed_indices)
+        dct.update(dict(doc_lists))
+        return list(dct.items())
+
     def docs_for_training_data(
         self, training_data: TrainingData
     ) -> Dict[Text, List[Any]]:
 
         attribute_docs = {}
         for attribute in SPACY_FEATURIZABLE_ATTRIBUTES:
-
             texts = [self.get_text(e, attribute) for e in training_data.intent_examples]
+            # Index and freeze indices of the training samples for preserving the order
+            # after processing the data.
+            freezed_indices = [(idx, text) for idx, text in enumerate(texts)]
+            # Currently, none existent attribute samples are None which leads to an error
+            # in spaCy and are therefore set as empty strings by Rasa. Since third-party
+            # libraries relying on spaCy might have problems with Doc-objects created on empty
+            # strings, we further treat them separately.
+            docs_to_pipe = list(filter(lambda fi_tpl: fi_tpl[1] != "", freezed_indices))
+            empty_docs = list(filter(lambda fi_tpl: fi_tpl[1] == "", freezed_indices))
+
+            # To preserve the order of the training samples, we send the non-empty samples to
+            # pipe and then simply update their original/index tuple.
+            docs = [
+                (idx_tpl[0], doc)
+                for idx_tpl, doc in zip(
+                    docs_to_pipe,
+                    [
+                        doc
+                        for doc in self.nlp.pipe(
+                            [txt for _, txt in docs_to_pipe], batch_size=50
+                        )
+                    ],
+                )
+            ]
+            # To avoid creating a Doc-object by parsing an empty string, we instantiate it directly
+            # via spaCy's API for every empty attribute sample and update its original/index tuple.
+            n_docs = [
+                (idx_tpl[0], doc)
+                for idx_tpl, doc in zip(
+                    empty_docs, [Doc(self.nlp.vocab) for doc in empty_docs]
+                )
+            ]
+
+            # After processing the training samples, we merge the resulting lists back into the
+            # tuple with our freezed indices.
+            attribute_document_list = self.merge_content_lists(
```

Comment:
I think the function has expanded too much. Can you break the three components into smaller functions for readability -
1. Filtering of texts
2. Fetching empty and non-empty docs
3. Merging docs(you already have this)

Reply:
Hi,

no problem - makes sense. I modified the code, added the test method, renamed most of the variables/methods such that they mostly tell a "story" of what happens.

Hope I got you right here! 

Regards